### ISSUE TITLE
microsoft account auth provider is twice as slow 

### ISSUE THREAD

**Issue by phgmacedo** [microsoft/vscode#156518]
Issue Type: <b>Performance Issue</b>

microsoft account auth provider is twice as slow as the github authentication provider

VS Code version: Code 1.69.2 (3b889b090b5ad5793f524b5d1d39fda662b96a2a, 2022-07-18T16:12:52.460Z)
OS version: Windows_NT x64 10.0.22000
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz (8 x 2304)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.79GB (7.75GB free)|
|Process Argv|--crash-reporter-id 53b331b0-53e4-47f4-ac88-511df9be7097|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (17)</summary>

Extension|Author (truncated)|Version
---|---|---
pythonsnippets3|Eri|3.3.14
gitlab-workflow|Git|3.47.2
tasks-statusbar|hed|0.1.0
magit|kah|0.6.29
git-graph|mhu|1.30.0
python|ms-|2022.10.1
vscode-pylance|ms-|2022.7.40
jupyter|ms-|2022.6.1201981810
jupyter-renderers|ms-|1.0.9
hexeditor|ms-|1.9.7
back-n-forth|nic|3.1.1
text-power-tools|qcz|1.34.0
vscode-xml|red|0.21.2022062916
vscode-yaml|red|1.9.1
datetime|rid|2.1.4
vscode-fileutils|sle|3.5.0
even-better-toml|tam|0.16.5

(1 theme extensions excluded)

</details>

**Comment by phgmacedo** [comment]
On startup, github authentication - signed in takes 30ms. The microsoft account signed out, shouldn't even be running, takes 97ms.
![gh](https://user-images.githubusercontent.com/92439290/181407282-9380dd36-34c7-4fb5-babc-fb7c21c8269d.png)
![ma](https://user-images.githubusercontent.com/92439290/181407293-64098dca-339d-4350-a9c4-9e8640ce214a.png)

**Comment by phgmacedo** [comment]
There are quite a few built-in extensions that load onStartupFinished but aren't exactly used, that I disable to improve the startup experience.

**Comment by phgmacedo** [comment]
Here`s a list of extensions I disabled, they save at least about 200ms of startup:
![disabled](https://user-images.githubusercontent.com/92439290/181408121-f303a989-61e9-4973-a49d-ce5f5b6c1bf4.png)

**Comment by TylerLeonhardt** [comment]
This is likely because of a few things:
* GitHub Account runs faster because specifically optimized it for github.dev/vscode.dev
* Microsoft Account runs slower because:
  * Microsoft sessions have refresh tokens and we refresh all tokens on startup to prevent the user getting logged out
  * We store each session it its own keyring account so more round trips with the OS

I can probably do a pass at following the same optimizations for Microsoft Authentication.

**Comment by phgmacedo** [comment]
> This is likely because of a few things:
> 
> * GitHub Account runs faster because specifically optimized it for github.dev/vscode.dev
> * Microsoft Account runs slower because:
>   
>   * Microsoft sessions have refresh tokens and we refresh all tokens on startup to prevent the user getting logged out
>   * We store each session it its own keyring account so more round trips with the OS
> 
> I can probably do a pass at following the same optimizations for Microsoft Authentication.

Yeah, but in this case, I’m not logged in into a Microsoft account. Therefore the extension does even need to be activated.

**Comment by TylerLeonhardt** [comment]
I think the problem there is that Settings Sync is activating both GitHub & Microsoft account extensions... @sandy081 do you think we can improve this? I know that Settings Sync stores the account that it uses somewhere. Maybe the logic could be:

*First time Settings Sync usage*
* Settings Sync looks in storage to see if account preference is set
* Nothing is there so do nothing until user wants to log in to Settings Sync
* User logs into Settings Sync and it **activates both extensions** to populate a quick pick to choose the account you wanna use
* User chooses either GH or one of the Microsoft accounts (since MS supports multi-account) and that is cached in Storage

*Return Settings Sync usage*
* Settings Sync looks in storage to see if account pref is set
* It is and it's GitHub great fetch the session from the GitHub Auth Provider
* Microsoft is never activated since it's not used by Settings Sync

**Comment by sandy081** [comment]
Yeah makes sense to me.
